pkg/cmd/server/origin/handlers.go

```diff
@@ -302,125 +302,119 @@ func assetServerRedirect(handler http.Handler, assetPublicURL string) http.Handl
 
 func (c *MasterConfig) impersonationFilter(handler http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		requestedSubject := req.Header.Get(authenticationapi.ImpersonateUserHeader)
-		if len(requestedSubject) == 0 {
+		requestedUser := req.Header.Get(authenticationapi.ImpersonateUserHeader)
+		if len(requestedUser) == 0 {
 			handler.ServeHTTP(w, req)
 			return
 		}
 
-		resource, namespace, name, err := parseRequestedSubject(requestedSubject)
-		if err != nil {
-			forbidden(err.Error(), nil, w, req)
-			return
-		}
+		subjects := authorizationapi.BuildSubjects([]string{requestedUser}, req.Header[authenticationapi.ImpersonateGroupHeader],
+			// validates whether the usernames are regular users or system users
+			uservalidation.ValidateUserName,
+			// validates group names are regular groups or system groups
+			uservalidation.ValidateGroupName)
 
 		ctx, exists := c.RequestContextMapper.Get(req)
 		if !exists {
 			forbidden("context not found", nil, w, req)
 			return
 		}
 
-		actingAsAttributes := &authorizer.DefaultAuthorizationAttributes{
-			Verb:         "impersonate",
-			APIGroup:     resource.Group,
-			Resource:     resource.Resource,
-			ResourceName: name,
-		}
-		authCheckCtx := kapi.WithNamespace(ctx, namespace)
-
-		allowed, reason, err := c.Authorizer.Authorize(authCheckCtx, actingAsAttributes)
-		if err != nil {
-			forbidden(err.Error(), actingAsAttributes, w, req)
-			return
-		}
-		if !allowed {
-			forbidden(reason, actingAsAttributes, w, req)
-			return
-		}
+		// if groups are not specified, then we need to look them up differently depending on the type of user
+		// if they are specified, then they are the authority
+		groupsSpecified := len(subjects) > 1
```

Comment:
> subjects might not be groups... shouldn't this be checking whether the header was specified?

In order to get more than one subject, you must have specified a group.  The `requestedUser` is guaranteed to produce one subject.

I can rewrite if you find the other comparison more obvious, but it is effectively identical.